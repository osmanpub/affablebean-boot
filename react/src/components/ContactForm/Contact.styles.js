import styled from "styled-components";

export const InfoBox = styled.div`
  margin-left: 150px;
  /*padding: 10px 25px 0 460px*/
`;
