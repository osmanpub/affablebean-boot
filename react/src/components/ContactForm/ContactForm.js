import React, { Component } from "react";
import { Link } from "react-router-dom";
import { sendFeedback } from "../../net/contact";
import { validateField } from "../../utils";

export class ContactForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: "",
      mag: "",
      name1: "",
      subjectId: ""
    };

    this.emailErrorRef = React.createRef();
    this.emailInputRef = React.createRef();

    this.msgErrorRef = React.createRef();
    this.msgInputRef = React.createRef();

    this.nameErrorRef = React.createRef();
    this.nameInputRef = React.createRef();

    this.subjectErrorRef = React.createRef();
    this.subjectInputRef = React.createRef();

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(event) {
    const input = event.target;
    this.setState({
      [input.name]: input.value
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    let validForm = true;

    if (!validateField(this.emailInputRef, this.emailErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.msgInputRef, this.msgErrorRef, 8, 256)) {
      validForm = false;
    }

    if (!validateField(this.nameInputRef, this.nameErrorRef, 8, 45)) {
      validForm = false;
    }

    if (!validateField(this.subjectInputRef, this.subjectErrorRef, 1, 1)) {
      validForm = false;
    }

    if (validForm) {
      sendFeedback({
        email: this.emailInputRef.current.value,
        msg: this.msgInputRef.current.value,
        name: this.nameInputRef.current.value,
        subjectId: this.subjectInputRef.current.value
      });

      this.emailInputRef.current.value = "";
      this.msgInputRef.current.value = "";
      this.nameInputRef.current.value = "";
      this.subjectInputRef.current.value = "";
    }
  }

  render() {
    const { subjects } = this.props;
    const subjectsList = subjects.items.map(subject => (
      <option key={subject._links.self.href} value={subject.id}>
        {subject.name}
      </option>
    ));

    return (
      <div className="singleColumn">
        <div>
          <h2>Contact Us Form</h2>
          <p>
            You can use this form for any comments or questions about our
            company or brands.
            <br />
            For general enquiries please call toll free on 1-800-BEANS-R-US
          </p>
        </div>
        <div>
          <p>(* Asterisks indicate required field)</p>
        </div>
        <form className="form-horizontal" onSubmit={this.handleSubmit}>
          <div className="form-group">
            <label htmlFor="subject_sel" className={`col-sm-2 control-label`}>
              subject
            </label>
            <div className="col-sm-10">
              <select
                ref={this.subjectInputRef}
                name="subject_sel"
                className="form-control"
              >
                {subjectsList}
              </select>
              <div className="formError" ref={this.subjectErrorRef}>
                Select a subject
              </div>
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="name" className={`col-sm-2 control-label`}>
              name
            </label>
            <div className="col-sm-10">
              <input
                ref={this.nameInputRef}
                type="text"
                className="form-control"
                name="name1"
                maxLength="45"
                placeholder="At least 8 chars and no more than 45 chars"
                size="31"
                onChange={this.handleChange}
                value={this.state.name}
              />
            </div>
            <div className="formError" ref={this.nameErrorRef}>
              Name shoud be at least 8 chars and no more than 45 chars
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="email" className={`col-sm-2 control-label`}>
              email
            </label>
            <div className="col-sm-10">
              <input
                ref={this.emailInputRef}
                type="email"
                className="form-control"
                name="email"
                maxLength="45"
                placeholder="At least 8 chars and no more than 45 chars"
                size="31"
                onChange={this.handleChange}
                value={this.state.email}
              />
            </div>
            <div className="formError" ref={this.emailErrorRef}>
              Email shoud be at least 8 chars and no more than 45 chars
            </div>
          </div>
          <div className="form-group">
            <label htmlFor="msg" className={`col-sm-2 control-label`}>
              message
            </label>
            <div className="col-sm-10">
              <textarea
                ref={this.msgInputRef}
                className="form-control"
                name="msg"
                cols="25"
                rows="5"
                onChange={this.handleChange}
                value={this.state.msgl}
              />
            </div>
            <div className="formError" ref={this.msgErrorRef}>
              Message shoud be at least 8 chars and no more than 256 chars
            </div>
          </div>
          <div className="form-group">
            <div className={`col-sm-offset-2 col-sm-10`}>
              <button type="submit" className={`btn btn-primary`}>
                send feedback
              </button>
            </div>
          </div>
        </form>
        <div>
          <br />
          <h3>Privacy</h3>
          <p>
            This website is owned and controlled by:
            <br />
            <br /> The Affable Bean Company Limited <br />
            Registered office: AffableBean House, Roast Drive, <br /> Pothead,
            POT123, BeanieLand
          </p>
          <p>(hereinafter referred to as "AffableBean")</p>

          <p>
            AffableBean is responsible for the processing of any information
            collected by or on behalf of AffableBean on or via this website. All
            such information will be processed in accordance with our&nbsp;
            <Link to="/privacy">privacy policy</Link>.
          </p>
        </div>
      </div>
    );
  }
}
