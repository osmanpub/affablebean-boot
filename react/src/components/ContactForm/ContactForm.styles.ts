export const styles = {
  toast: {
    backgroundColor: "green",
    color: "white",
    fontWeight: "bold"
  }
};
