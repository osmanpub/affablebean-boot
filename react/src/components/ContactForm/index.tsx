import React from "react";
// import { types, useAlert } from "react-alert";
import { useForm } from "react-hook-form";
import { connect, useDispatch } from "react-redux";
import { Link } from "react-router-dom";
// import { Link, useHistory } from "react-router-dom";
import { getId } from "../../helpers/utils";
import { Subjects, SubjectState } from "../../interfaces/subjects";
import { FormErrors } from "../../interfaces/ui";
import { sendFeedback } from "../../net/contact";
import { RootState } from "../../redux";
import { goHome, setFormErrors } from "../../redux/ui";

type Props = {
  formErrors: Array<FormErrors>;
  goHome: Function;
  home: boolean;
  setFormErrors: Function;
  subjects: Subjects;
};

type FormData = {
  email: string;
  name: string;
  msg: string;
  subject: string;
};

function ContactForm(props: Props) {
  const { formErrors, goHome, home, setFormErrors, subjects } = props;
  // Problem with jest - https://github.com/schiehll/react-alert/issues/148
  // const alert = useAlert();
  const dispatch = useDispatch();
  // const history = useHistory();
  const { register, handleSubmit, errors } = useForm<FormData>();

  const onSubmit = handleSubmit(({ email, msg, name, subject }) => {
    dispatch(sendFeedback({ email, msg, name, subjectId: subject }));
  });

  const subjectsList = subjects.items.map((subject: SubjectState) => {
    const id = getId(subject);
    return (
      <option key={id} value={id}>
        {subject.name}
      </option>
    );
  });

  if (formErrors && Array.isArray(formErrors) && formErrors.length > 0) {
    let msg = "";

    formErrors.forEach(error => {
      msg += `Field "${error.param}" with value "${error.value}" has the following problem:\n"${error.msg}"`;
    });

    setFormErrors([]);
    alert(
      `There was a problem saving your message.\nPlease correct the following errors:\n${msg}`
    );

    // alert.show(
    //   `There was a problem saving your message.\nPlease correct the following errors:\n${msg}`,
    //   {
    //     timeout: 0,
    //     type: types.ERROR
    //   }
    // );
  } else if (home) {
    goHome(false);
    alert("Message sent successfully!");

    // alert.show("Message sent successfully!", {
    //   onClose: () => history.push("/"),
    //   timeout: 3000,
    //   type: types.SUCCESS
    // });
  }

  return (
    <div className="singleColumn">
      <div>
        <h2 data-cy="contact-intro">Contact Us Form</h2>
        <p>
          You can use this form for any comments or questions about our company
          or brands.
          <br />
          For general enquiries please call toll free on 1-800-BEANS-R-US
        </p>
      </div>
      <div>
        <p>(* Asterisks indicate required field)</p>
      </div>
      <form className="form-horizontal" onSubmit={onSubmit}>
        <div className="form-group">
          <label htmlFor="subject" className={`col-sm-2 control-label`}>
            subject
          </label>
          <div className="col-sm-10">
            <select
              className="form-control"
              ref={register({ required: true })}
              name="subject"
            >
              {subjectsList}
            </select>
          </div>
          {errors.subject && <div className="formError">Select a subject</div>}
        </div>
        <div className="form-group">
          <label htmlFor="name" className={`col-sm-2 control-label`}>
            name
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="contact-name"
              name="name"
              placeholder="At least 3 chars and no more than 64 chars"
              size={31}
              ref={register({ required: true, minLength: 3, maxLength: 64 })}
              type="text"
            />
          </div>
          {errors.name && (
            <div className="formError">
              Name shoud be at least 3 chars and no more than 64 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="email" className={`col-sm-2 control-label`}>
            email
          </label>
          <div className="col-sm-10">
            <input
              className="form-control"
              data-cy="contact-email"
              name="email"
              placeholder="At least 8 chars and no more than 32 chars"
              size={31}
              ref={register({ required: true, minLength: 8, maxLength: 32 })}
              type="email"
            />
          </div>
          {errors.email && (
            <div className="formError">
              Email shoud be at least 8 chars and no more than 32 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <label htmlFor="msg" className={`col-sm-2 control-label`}>
            message
          </label>
          <div className="col-sm-10">
            <textarea
              className="form-control"
              cols={25}
              rows={5}
              data-cy="contact-msg"
              name="msg"
              placeholder="At least 8 chars and no more than 1024 chars"
              ref={register({ required: true, minLength: 8, maxLength: 1024 })}
            />
          </div>
          {errors.msg && (
            <div className="formError">
              Message shoud be at least 8 chars and no more than 1024 chars
            </div>
          )}
        </div>
        <div className="form-group">
          <div className={`col-sm-offset-2 col-sm-10`}>
            <button
              className={`btn btn-primary`}
              data-cy="contact-submit"
              type="submit"
            >
              send feedback
            </button>
          </div>
        </div>
      </form>
      <div>
        <br />
        <h3>Privacy</h3>
        <p>
          This website is owned and controlled by:
          <br />
          <br /> The Affable Bean Company Limited <br />
          Registered office: AffableBean House, Roast Drive, <br /> Pothead,
          POT123, BeanieLand
        </p>
        <p>(hereinafter referred to as "AffableBean")</p>

        <p>
          AffableBean is responsible for the processing of any information
          collected by or on behalf of AffableBean on or via this website. All
          such information will be processed in accordance with our&nbsp;
          <Link to="/privacy">privacy policy</Link>.
        </p>
      </div>
    </div>
  );
}

const mapStateToProps = (state: RootState) => ({
  formErrors: state.ui.formErrors,
  home: state.ui.home,
  subjects: state.subjects
});

const mapDispatchToProps = {
  goHome,
  setFormErrors
};

export default connect(mapStateToProps, mapDispatchToProps)(ContactForm);
