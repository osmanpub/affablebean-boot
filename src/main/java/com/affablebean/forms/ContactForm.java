package com.affablebean.forms;

import javax.validation.constraints.Email;
import javax.validation.constraints.Max;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

public class ContactForm {

	@NotNull
	@Min(1)
	@Max(7)
	private Integer subjectId;

	@NotNull
	@Email
	@Size(min = 1, max = 45)
	private String email;

	@NotNull
	@Size(min = 1, max = 1000)
	private String msg;

	@NotNull
	@Size(min = 1, max = 45)
	private String name;

	public Integer getSubjectId() {
		return subjectId;
	}

	public void setSubjectId(Integer subjectId) {
		this.subjectId = subjectId;
	}

	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getMsg() {
		return msg;
	}

	public void setMsg(String msg) {
		this.msg = msg;
	}

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	@Override
	public String toString() {
		return "ContactForm [subjectId=" + subjectId + ", email=" + email + ", msg=" + msg + ", name=" + name + "]";
	}

}
