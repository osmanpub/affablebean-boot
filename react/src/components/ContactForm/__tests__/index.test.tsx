import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import ContactForm from "..";

const mockStore = configureStore([]);

const initialState = {
  cart: {
    didInvalidate: false,
    isFetching: false,
    items: [],
    numberOfItems: 0,
    subtotal: 0
  },
  subjects: {
    didInvalidate: false,
    isFetching: false,
    items: [
      {
        id: 1,
        name: "Brands or product"
      },
      {
        id: 2,
        name: "Investor relations"
      },
      {
        id: 3,
        name: "Sustainability"
      },
      {
        id: 4,
        name: "The Company"
      },
      {
        id: 5,
        name: "Media enquiry"
      },
      {
        id: 6,
        name: "Website feedback"
      },
      {
        id: 7,
        name: "Other"
      }
    ]
  },
  ui: {
    formErrors: []
  }
};

const store = mockStore(initialState);

describe("<ContactForm />", () => {
  it("renders correctly", () => {
    const component = renderer
      .create(
        <Router>
          <Provider store={store}>
            <ContactForm />
          </Provider>
        </Router>
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});
