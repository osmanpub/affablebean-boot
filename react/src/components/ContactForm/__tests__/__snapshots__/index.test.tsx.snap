// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> renders correctly 1`] = `
<div
  className="singleColumn"
>
  <div>
    <h2
      data-cy="contact-intro"
    >
      Contact Us Form
    </h2>
    <p>
      You can use this form for any comments or questions about our company or brands.
      <br />
      For general enquiries please call toll free on 1-800-BEANS-R-US
    </p>
  </div>
  <div>
    <p>
      (* Asterisks indicate required field)
    </p>
  </div>
  <form
    className="form-horizontal"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="subject"
      >
        subject
      </label>
      <div
        className="col-sm-10"
      >
        <select
          className="form-control"
          name="subject"
        >
          <option
            value={1}
          >
            Brands or product
          </option>
          <option
            value={2}
          >
            Investor relations
          </option>
          <option
            value={3}
          >
            Sustainability
          </option>
          <option
            value={4}
          >
            The Company
          </option>
          <option
            value={5}
          >
            Media enquiry
          </option>
          <option
            value={6}
          >
            Website feedback
          </option>
          <option
            value={7}
          >
            Other
          </option>
        </select>
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="name"
      >
        name
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          data-cy="contact-name"
          name="name"
          placeholder="At least 3 chars and no more than 64 chars"
          size={31}
          type="text"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="email"
      >
        email
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          data-cy="contact-email"
          name="email"
          placeholder="At least 8 chars and no more than 32 chars"
          size={31}
          type="email"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="msg"
      >
        message
      </label>
      <div
        className="col-sm-10"
      >
        <textarea
          className="form-control"
          cols={25}
          data-cy="contact-msg"
          name="msg"
          placeholder="At least 8 chars and no more than 1024 chars"
          rows={5}
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="col-sm-offset-2 col-sm-10"
      >
        <button
          className="btn btn-primary"
          data-cy="contact-submit"
          type="submit"
        >
          send feedback
        </button>
      </div>
    </div>
  </form>
  <div>
    <br />
    <h3>
      Privacy
    </h3>
    <p>
      This website is owned and controlled by:
      <br />
      <br />
       The Affable Bean Company Limited 
      <br />
      Registered office: AffableBean House, Roast Drive, 
      <br />
       Pothead, POT123, BeanieLand
    </p>
    <p>
      (hereinafter referred to as "AffableBean")
    </p>
    <p>
      AffableBean is responsible for the processing of any information collected by or on behalf of AffableBean on or via this website. All such information will be processed in accordance with ourÂ 
      <a
        href="/privacy"
        onClick={[Function]}
      >
        privacy policy
      </a>
      .
    </p>
  </div>
</div>
`;
